File: from-3101-to-3200/3160-find-the-number-of-distinct-colors-among-the-balls/solution.go
func queryResults(limit int, queries [][]int) []int {
    colorMap := make(map[int]int)
    colorCount := make(map[int]int)
    result := make([]int, len(queries))
    numColors := 0

    for i, q := range queries {
        ball, newColor := q[0], q[1]
        if oldColor, ok := colorMap[ball]; ok {
            colorCount[oldColor] -= 1
            if colorCount[oldColor] == 0 {
                numColors -= 1
            }
        }

        colorMap[ball] = newColor
        colorCount[newColor] += 1
        if colorCount[newColor] == 1 {
            numColors += 1
        }

        result[i] = numColors
    }
    return result
}

File: from-3101-to-3200/3160-find-the-number-of-distinct-colors-among-the-balls/solution.py
class Solution:
    def queryResults(self, limit: int, queries: List[List[int]]) -> List[int]:
        ball_map = {}
        colors = defaultdict(int)
        result = [0] * len(queries)

        for i, query in enumerate(queries):
            ball, color = query
            if ball in ball_map:
                balance = ball_map[ball]
                if colors[balance] == 1:
                    del colors[balance]
                else:
                    colors[balance] -= 1
            
            ball_map[ball] = color
            colors[color] += 1
            result[i] = len(colors)

        return result
                